components/eam/src/physics/cam/zm_microphysics.F90

```diff
@@ -0,0 +1,3337 @@
+                 npccn(k) = max(0._r8,npccn(k))
+                 ncmax = dum2l(i,k)
+              else
+                 npccn(k)=0._r8
+                 ncmax = 0._r8
+              end if
+
+              !..............................................................................
+              !ice nucleation
+              es(i,k)  = svp_water(t(i,k))     ! over water in mixed clouds
+              esi(i,k) = svp_ice(t(i,k))     ! over ice
+              qs(i,k) = 0.622_r8*es(i,k)/(ph(i,k) - (1.0_r8-0.622_r8)*es(i,k))
+              qs(i,k) = min(1.0_r8,qs(i,k))
+              if (qs(i,k) < 0.0_r8)  qs(i,k) = 1.0_r8
+
+              relhum(i,k)= 1.0_r8
+
+              if (t(i,k).lt.tmelt ) then
+
+                 ! compute aerosol number for so4, soot, and dust with units #/cm^3
+                 so4_num  = 0._r8
+                 soot_num = 0._r8
+                 dst1_num = 0._r8
+                 dst2_num = 0._r8
+                 dst3_num = 0._r8
+                 dst4_num = 0._r8
+
+                 if (aero%scheme == 'modal') then          
+
+                    !For modal aerosols, assume for the upper troposphere:
+                    ! soot = accumulation mode
+                    ! sulfate = aiken mode
+                    ! dust = coarse mode
+                    ! since modal has internal mixtures.
+                    soot_num = 0.5_r8*(aero%numg_a(i,k-1,aero%mode_accum_idx)             &
+                                      +aero%numg_a(i,k,aero%mode_accum_idx))*rho(i,k)*1.0e-6_r8
+                    dmc  = 0.5_r8*(aero%mmrg_a(i,k-1,aero%coarse_dust_idx,aero%mode_coarse_idx)      &
+                                  +aero%mmrg_a(i,k,aero%coarse_dust_idx,aero%mode_coarse_idx))
+                    ssmc = 0.5_r8*(aero%mmrg_a(i,k-1,aero%coarse_nacl_idx,aero%mode_coarse_idx)      &
+                                  +aero%mmrg_a(i,k,aero%coarse_nacl_idx,aero%mode_coarse_idx))  
+                    if (dmc > 0._r8) then
+                        dst_num = dmc/(ssmc + dmc) *(aero%numg_a(i,k-1,aero%mode_coarse_idx)         & 
+                                  + aero%numg_a(i,k,aero%mode_coarse_idx))*0.5_r8*rho(i,k)*1.0e-6_r8
+                    else 
+                       dst_num = 0.0_r8
+                    end if
+                    dgnum_aitken = 0.5_r8*(aero%dgnumg(i,k,aero%mode_aitken_idx)+   &
+                                           aero%dgnumg(i,k-1,aero%mode_aitken_idx))     
+                    if (dgnum_aitken > 0._r8) then
+                       ! only allow so4 with D>0.1 um in ice nucleation
+                       so4_num  = 0.5_r8*(aero%numg_a(i,k-1,aero%mode_aitken_idx)+          &
+                          aero%numg_a(i,k,aero%mode_aitken_idx))*rho(i,k)*1.0e-6_r8         &
+                          * (0.5_r8 - 0.5_r8*erf(log(0.1e-6_r8/dgnum_aitken)/  &
+                          (2._r8**0.5_r8*log(aero%sigmag_aitken))))
+                    else 
+                       so4_num = 0.0_r8 
```

Comment:
@crterai @hwangacme I think there are two issues here. 

1. It misses the aerosol species in MAM4 and related flags. You can see nucleate_ice_cam.F90 to get some ideas.
In default MAM4, there are also sulfate, BC, POA, MOA, SOA aerosols in the coarse mode. It should be something like:

#if (defined MODAL_AERO_4MODE_MOM && defined RAIN_EVAP_TO_COARSE_AERO )
                     wght = dmc/(ssmc + dmc + so4mc + bcmc + pommc + soamc + mommc)
#elif (defined MODAL_AERO_4MODE_MOM)
                     wght = dmc/(ssmc + dmc + so4mc + mommc)
#elif (defined RAIN_EVAP_TO_COARSE_AERO) 
                     wght = dmc/(ssmc + dmc + so4mc + bcmc + pommc + soamc)
#else
                     wght = dmc/(ssmc + dmc) ! to keep FC5 binary identical 
#endif

This also need to make related changes for index of aerosol species (e.g., coarse_bc_idx, coarse_pom_idx), corresponding to the place where @hwangacme raised concerns.


2. The second issue, which is not that urgent, is that we need to modify the code here to consider species introduced by MOSAIC.

Reply:
Thanks, @mingxuanwupnnl. I think we will leave 2. to a separate PR in the future. Xiaoliang Song is looking into the code changes for 1.

Reply:
The code has been updated, thanks to Xiaoliang's efforts to include the necessary linkages for 1.

Reply:
@hwangacme @mingxuanwupnnl @crterai I want to add the stratospheric sulfate (mode 5) into consideration. In the upper troposphere, there are more sulfate in the mode 5 instead of coarse mode (mode 3), if MAM5 is activated. I can work with @mingxuanwupnnl to fix this issue in the future.

Reply:
@hwangacme @mingxuanwupnnl @crterai please discard my comments. I find this process is for dust nucleation. There is no dust in mode-5, so that it is not necessary to add mode-5 in this process.